File: koladata/operators/comparison.cc
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
#include "koladata/operators/comparison.h"

#include <type_traits>

#include "absl/log/check.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "absl/types/span.h"
#include "arolla/qexpr/operators/core/logic_operators.h"
#include "arolla/util/bytes.h"
#include "arolla/util/text.h"
#include "arolla/util/unit.h"
#include "koladata/data_bag.h"
#include "koladata/data_slice.h"
#include "koladata/data_slice_op.h"
#include "koladata/data_slice_qtype.h"
#include "koladata/internal/data_item.h"
#include "koladata/internal/dtype.h"
#include "koladata/internal/op_utils/equal.h"
#include "koladata/operators/binary_op.h"
#include "koladata/operators/utils.h"
#include "koladata/schema_utils.h"
#include "arolla/util/status_macros_backport.h"

namespace koladata::ops {

namespace {

// Similar to RETURN_IF_ERROR, but doesn't add an extra source location
// to the status.
#define KD_RETURN_AS_IS_IF_ERROR(X)      \
  if (absl::Status st = (X); !st.ok()) { \
    return st;                           \
  }

struct CanBeOrdered {
  template <class T1, class T2>
  constexpr static bool kIsInvocable =
      (std::is_arithmetic_v<T1> && std::is_arithmetic_v<T2> &&
      !std::is_same_v<T1, bool> && !std::is_same_v<T2, bool>) ||
      (std::is_same_v<T1, bool> && std::is_same_v<T2, bool>) ||
      (std::is_same_v<T1, arolla::Unit> && std::is_same_v<T2, arolla::Unit>) ||
      (std::is_same_v<T1, arolla::Text> && std::is_same_v<T2, arolla::Text>) ||
      (std::is_same_v<T1, arolla::Bytes> && std::is_same_v<T2, arolla::Bytes>);

  explicit constexpr CanBeOrdered(absl::string_view name1,
                                  absl::string_view name2)
      : name1(name1), name2(name2) {}

  absl::Status CheckArgs(const DataSlice& ds1, const DataSlice& ds2) const {
    KD_RETURN_AS_IS_IF_ERROR(ExpectCanBeOrdered(name1, ds1));
    KD_RETURN_AS_IS_IF_ERROR(ExpectCanBeOrdered(name2, ds2));
    return ExpectHaveCommonPrimitiveSchema({name1, name2}, ds1, ds2);
  }

  const absl::string_view name1, name2;
};

#undef KD_RETURN_AS_IS_IF_ERROR

}  // namespace

absl::StatusOr<DataSlice> Less(const DataSlice& x, const DataSlice& y) {
  return BinaryOpEval<arolla::MaskLessOp>(x, y, CanBeOrdered("x", "y"),
                                          BinaryOpReturns(schema::kMask));
}

absl::StatusOr<DataSlice> Greater(const DataSlice& x, const DataSlice& y) {
  return BinaryOpEval<arolla::MaskLessOp>(y, x, CanBeOrdered("y", "x"),
                                          BinaryOpReturns(schema::kMask));
}

absl::StatusOr<DataSlice> LessEqual(const DataSlice& x, const DataSlice& y) {
  return BinaryOpEval<arolla::MaskLessEqualOp>(x, y, CanBeOrdered("x", "y"),
                                               BinaryOpReturns(schema::kMask));
}

absl::StatusOr<DataSlice> GreaterEqual(const DataSlice& x, const DataSlice& y) {
  return BinaryOpEval<arolla::MaskLessEqualOp>(y, x, CanBeOrdered("y", "x"),
                                               BinaryOpReturns(schema::kMask));
}

absl::StatusOr<DataSlice> Equal(const DataSlice& x, const DataSlice& y) {
  // NOTE: Casting is handled internally by EqualOp. The schema compatibility is
  // still verified to ensure that e.g. ITEMID and OBJECT are not compared.
  RETURN_IF_ERROR(ExpectHaveCommonSchema({"x", "y"}, x, y));
  return DataSliceOp<internal::EqualOp>()(
      x, y, internal::DataItem(schema::kMask), nullptr);
}

}  // namespace koladata::ops

File: py/koladata/operators/comparison.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Comparison DataSlice operators."""

from arolla import arolla as _arolla
from koladata.operators import masking as _masking
from koladata.operators import op_repr as _op_repr
from koladata.operators import optools as _optools
from koladata.operators import qtype_utils as _qtype_utils

_P = _arolla.P


@_optools.add_to_registry(
    aliases=['kd.equal'],
    repr_fn=_op_repr.equal_repr,
    via_cc_operator_package=True,
)
@_optools.as_backend_operator(
    'kd.comparison.equal',
    qtype_constraints=[
        _qtype_utils.expect_data_slice(_P.x),
        _qtype_utils.expect_data_slice(_P.y),
    ],
)
def equal(x, y):  # pylint: disable=unused-argument
  """Returns present iff `x` and `y` are equal.

  Pointwise operator which takes a DataSlice and returns a MASK indicating
  iff `x` and `y` are equal. Returns `kd.present` for equal items and
  `kd.missing` in other cases.

  Args:
    x: DataSlice.
    y: DataSlice.
  """
  raise NotImplementedError('implemented in the backend')


@_optools.add_to_registry(
    aliases=['kd.full_equal'], via_cc_operator_package=True
)
@_optools.as_lambda_operator(
    'kd.comparison.full_equal',
    qtype_constraints=[
        _qtype_utils.expect_data_slice(_P.x),
        _qtype_utils.expect_data_slice(_P.y),
    ],
)
def full_equal(x, y):  # pylint: disable=unused-argument
  """Returns present iff all present items in `x` and `y` are equal.

  The result is a zero-dimensional DataItem. Note that it is different from
  `kd.all(x == y)`.

  For example,
    kd.full_equal(kd.slice([1, 2, 3]), kd.slice([1, 2, 3])) -> kd.present
    kd.full_equal(kd.slice([1, 2, 3]), kd.slice([1, 2, None])) -> kd.missing
    kd.full_equal(kd.slice([1, 2, None]), kd.slice([1, 2, None])) -> kd.present

  Args:
    x: DataSlice.
    y: DataSlice.
  """
  return _masking.all_((x == y) | (_masking.has_not(x) & _masking.has_not(y)))


@_optools.add_to_registry(
    aliases=['kd.not_equal'],
    repr_fn=_op_repr.not_equal_repr,
    via_cc_operator_package=True,
)
@_optools.as_lambda_operator(
    'kd.comparison.not_equal',
    qtype_constraints=[
        _qtype_utils.expect_data_slice(_P.x),
        _qtype_utils.expect_data_slice(_P.y),
    ],
)
def not_equal(x, y):
  """Returns present iff `x` and `y` are not equal.

  Pointwise operator which takes a DataSlice and returns a MASK indicating
  iff `x` and `y` are not equal. Returns `kd.present` for not equal items and
  `kd.missing` in other cases.

  Args:
    x: DataSlice.
    y: DataSlice.
  """
  return ~(x == y) & (_masking.has(x) & _masking.has(y))


@_optools.add_to_registry(
    aliases=['kd.greater'],
    repr_fn=_op_repr.greater_repr,
    via_cc_operator_package=True,
)
@_optools.as_backend_operator(
    'kd.comparison.greater',
    qtype_constraints=[
        _qtype_utils.expect_data_slice(_P.x),
        _qtype_utils.expect_data_slice(_P.y),
    ],
)
def greater(x, y):  # pylint: disable=unused-argument
  """Returns present iff `x` is greater than `y`.

  Pointwise operator which takes a DataSlice and returns a MASK indicating
  iff `x` is greater than `y`. Returns `kd.present` when `x` is greater and
  `kd.missing` when `x` is less than or equal to `y`.

  Args:
    x: DataSlice.
    y: DataSlice.
  """
  raise NotImplementedError('implemented in the backend')


@_optools.add_to_registry(
    aliases=['kd.less'],
    repr_fn=_op_repr.less_repr,
    via_cc_operator_package=True,
)
@_optools.as_backend_operator(
    'kd.comparison.less',
    qtype_constraints=[
        _qtype_utils.expect_data_slice(_P.x),
        _qtype_utils.expect_data_slice(_P.y),
    ],
)
def less(x, y):  # pylint: disable=unused-argument
  """Returns present iff `x` is less than `y`.

  Pointwise operator which takes a DataSlice and returns a MASK indicating
  iff `x` is less than `y`. Returns `kd.present` when `x` is less and
  `kd.missing` when `x` is greater than or equal to `y`.

  Args:
    x: DataSlice.
    y: DataSlice.
  """
  raise NotImplementedError('implemented in the backend')


@_optools.add_to_registry(
    aliases=['kd.greater_equal'],
    repr_fn=_op_repr.greater_equal_repr,
    via_cc_operator_package=True,
)
@_optools.as_backend_operator(
    'kd.comparison.greater_equal',
    qtype_constraints=[
        _qtype_utils.expect_data_slice(_P.x),
        _qtype_utils.expect_data_slice(_P.y),
    ],
)
def greater_equal(x, y):  # pylint: disable=unused-argument
  """Returns present iff `x` is greater than or equal to `y`.

  Pointwise operator which takes a DataSlice and returns a MASK indicating
  iff `x` is greater than or equal to `y`. Returns `kd.present` when `x` is
  greater than or equal to `y` and `kd.missing` when `x` is less than `y`.

  Args:
    x: DataSlice.
    y: DataSlice.
  """
  raise NotImplementedError('implemented in the backend')


@_optools.add_to_registry(
    aliases=['kd.less_equal'],
    repr_fn=_op_repr.less_equal_repr,
    via_cc_operator_package=True,
)
@_optools.as_backend_operator(
    'kd.comparison.less_equal',
    qtype_constraints=[
        _qtype_utils.expect_data_slice(_P.x),
        _qtype_utils.expect_data_slice(_P.y),
    ],
)
def less_equal(x, y):  # pylint: disable=unused-argument
  """Returns present iff `x` is less than or equal to `y`.

  Pointwise operator which takes a DataSlice and returns a MASK indicating
  iff `x` is less than or equal to `y`. Returns `kd.present` when `x` is
  less than or equal to `y` and `kd.missing` when `x` is greater than `y`.

  Args:
    x: DataSlice.
    y: DataSlice.
  """
  raise NotImplementedError('implemented in the backend')
